### PR TITLE
Added support for Create type strings for array/boolean/int/float types.

### LINKED ISSUE #87: Create type strings for array/boolean/int/float types
**Description:**
Implement type strings for int/float/boolean and Array types.
Create type strings(mangled type names) and send to Rust Functions.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### LINKED ISSUE #88: Migrate array cast to use generated type string
**Description:**
below should call now rust
```
int[] i = [3];
any ix = i;
int[] j = (int[]) ix;
// assert j  == i
```
following should abort

```
int[] i = 3;
any ix = i;
float[] j = (float[]) ix;
```

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**
Depends on #87

### PR DESCRIPTION
**Pull request by ramakotareddy**
## Purpose
-> Added support for Create type strings for array/boolean/int/float types #87 .
-> Changes related to Migrate array cast to use generated type string #88 . 

Resolves #87
Resolves #88

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/llvmArrayTypeInsn.cpp [hunk 1]
```diff
@@ -27,13 +28,15 @@ void ArrayInsn::translate(LLVMModuleRef &modRef) {
   LLVMValueRef lhsOpRef = funcObj->getLocalVarRefUsingId(lhsName);
   if (!lhsOpRef)
     lhsOpRef = pkgObj->getGlobalVarRefUsingId(lhsName);
```
Comment (ruvi-d) [inline compiler/llvmArrayTypeInsn.cpp:30]: The lhs generated type is `i32` for arrays. as in 
```
%"%0" = alloca i32, align 4
 %0 = call i32 @new_int_array(i32 %"%3_temp")
```
Shouldn't %0 be of type `i8*` to capture the opaque pointer returned from the Rust runtime?

File: compiler/llvmTypeCastInsn.cpp [hunk 2]
```diff
@@ -92,3 +132,75 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
...
+  if (!addedFuncRef)
+    addedFuncRef = getIsSameTypeDeclaration(modRef, pkgObj, lhsRef, rhsRef);
+  paramRefs[0] = lhsRef;
+  paramRefs[1] = rhsRef;
+  return LLVMBuildCall(builder, addedFuncRef, paramRefs, 2, "call");
+}
+
+char const *TypeCastInsn::typeStringMangleName(LLVMValueRef typeVal,
+                                               TypeTagEnum typeTag,
+                                               BIRFunction *funcObj) {
```
Comment (ruvi-d) [inline compiler/llvmTypeCastInsn.cpp:166]: `funcObj` is not required?

File: compiler/llvmTypeCastInsn.cpp [hunk 3]
```diff
@@ -92,3 +132,75 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
...
+  char const *finalString;
+  switch (typeTag) {
```
Comment (ruvi-d) [inline compiler/llvmTypeCastInsn.cpp:168]: Can you point to the tests that cover these switch cases?

File: compiler/llvmTypeCastInsn.cpp [hunk 4]
```diff
@@ -92,3 +132,75 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
...
+  LLVMValueRef *paramRefs = new LLVMValueRef[2];
+  LLVMValueRef addedFuncRef = pkgObj->getFunctionRefBasedOnName("is_same_type");
+  if (!addedFuncRef)
+    addedFuncRef = getIsSameTypeDeclaration(modRef, pkgObj, lhsRef, rhsRef);
+  paramRefs[0] = lhsRef;
+  paramRefs[1] = rhsRef;
+  return LLVMBuildCall(builder, addedFuncRef, paramRefs, 2, "call");
+}
+
+char const *TypeCastInsn::typeStringMangleName(LLVMValueRef typeVal,
```
Comment (ruvi-d) [inline compiler/llvmTypeCastInsn.cpp:164]: Isn't it dangerous to return a pointer from within a function stack?

**Review by ruvi-d — CHANGES_REQUESTED** [review]
Please respond to the review comments as soon as possible

**Comment by ramakotareddy** [comment]
Thanks for your comments Ruvi. I will fix.

File: compiler/BIR.h [hunk 5]
```diff
@@ -807,6 +813,8 @@ class BIRFunction : public BIRNode {
   LLVMTypeRef getLLVMFuncRetTypeRefOfType(VarDecl *vDecl);
   VarDecl *getNameVarDecl(string opName);
   const char *getTypeNameOfTypeTag(TypeTagEnum typeTag);
+  LLVMValueRef generateAbortInsn (LLVMBuilderRef &builder,
+                LLVMModuleRef &modRef);
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/BIR.h:817]: Fixed

File: compiler/BIR.h [hunk 6]
```diff
@@ -405,6 +405,12 @@ class TypeCastInsn : public NonTerminatorInsn {
   void setTypeDecl(TypeDecl *tDecl) { typeDecl = tDecl; }
   void setTypesChecking(bool checktypes) { checkTypes = checktypes; }
   void translate(LLVMModuleRef &modRef);
+  LLVMValueRef getIsSameTypeDeclaration (LLVMModuleRef &modRef,
+                BIRPackage *pkg, LLVMValueRef lhsRef, LLVMValueRef rhsRef);
+  LLVMValueRef isSameType(LLVMModuleRef &modRef, LLVMValueRef lhsRef,
+                LLVMValueRef rhsRef);
+  char const* typeStringMangleName(LLVMValueRef typeVal, TypeTagEnum typeTag,
+                                BIRFunction *funcObj);
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/BIR.h:413]: Fixed

File: compiler/llvmFunctionGen.cpp [hunk 7]
```diff
@@ -147,6 +162,71 @@ void BIRFunction::translateFunctionBody(LLVMModuleRef &modRef) {
...
   }
+
+  // Splitting Basicblock after 'is_same_type()' function call and
+  // based on is_same_type() function result, crating branch condition using
+  // New Basicblock (ifBB) and abortBB(elseBB).
+  // In IfBB we are doing casing and from ElseBB Aborting.
+
+  unsigned totalBB = LLVMCountBasicBlocks(getNewFunctionRef());
+  LLVMBasicBlockRef currentBB = LLVMGetFirstBasicBlock(getNewFunctionRef());
+  ;
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmFunctionGen.cpp:173]: Fixed

File: compiler/llvmFunctionGen.cpp [hunk 8]
```diff
@@ -147,6 +162,71 @@ void BIRFunction::translateFunctionBody(LLVMModuleRef &modRef) {
     LLVMPositionBuilderAtEnd(builder, bb->getLLVMBBRef());
     bb->translate(modRef);
   }
+
+  // Splitting Basicblock after 'is_same_type()' function call and
+  // based on is_same_type() function result, crating branch condition using
+  // New Basicblock (ifBB) and abortBB(elseBB).
+  // In IfBB we are doing casing and from ElseBB Aborting.
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmFunctionGen.cpp:169]: Splitted this logic into separate function.

File: compiler/llvmFunctionGen.cpp [hunk 9]
```diff
@@ -98,6 +98,20 @@ LLVMTypeRef BIRFunction::getLLVMFuncRetTypeRefOfType(VarDecl *vDecl) {
   }
 }
 
+LLVMValueRef BIRFunction::generateAbortInsn(LLVMBuilderRef &builder,
+                                            LLVMModuleRef &modRef) {
+  BIRPackage *pkgObj = getPkgAddress();
+  LLVMValueRef addedFuncRef = pkgObj->getFunctionRefBasedOnName("abort");
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmFunctionGen.cpp:104]: @ruvi-d , here 'abort' string we are using to get the FunctionRef from Map like "getFunctionRefBasedOnName("new_int_array")"  and "getFunctionRefBasedOnName("map_store_int")".
Can you please give more details on magic string here.

File: compiler/llvmFunctionGen.cpp [hunk 10]
```diff
@@ -147,6 +162,71 @@ void BIRFunction::translateFunctionBody(LLVMModuleRef &modRef) {
...
+          // Inserting elseLLVMBB (abort BB) after New (bb0.split) basicblock.
+          New->getParent()->getBasicBlockList().insertAfter(New->getIterator(),
+                                                            unwrap(elseLLVMBB));
+          currentBB = LLVMGetNextBasicBlock(currentBB);
+          break;
+        }
+      }
+    }
+    currentBB = LLVMGetNextBasicBlock(currentBB);
+  }
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmFunctionGen.cpp:229]: Fixed

File: compiler/llvmTypeCastInsn.cpp (hunk 3 above)
Comment (ramakotareddy) [reply compiler/llvmTypeCastInsn.cpp:168]: anyTypeParam.bal
 boolToAnyTypeCast.bal
 floatTypecast.bal.skip
 multipleAnyTypeParams.bal
 stringTypeCast.bal

Above tests (inside test/casting dir) will cover these switch cases.

File: compiler/llvmTypeCastInsn.cpp (hunk 2 above)
Comment (ramakotareddy) [reply compiler/llvmTypeCastInsn.cpp:166]: Removed funcObj

Comment (ramakotareddy) [reply compiler/llvmTypeCastInsn.cpp:166]: Fixed

File: compiler/llvmFunctionGen.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (ruvi-d) [reply compiler/llvmFunctionGen.cpp:104]: Please refer this: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-magic

File: compiler/llvmPackageGen.cpp [hunk 11]
```diff
@@ -27,6 +27,16 @@ LLVMValueRef BIRPackage::getGlobalVarRefUsingId(string globVar) {
 void BIRPackage::translate(LLVMModuleRef &modRef) {
   // String Table initialization
   strBuilder = new StringTableBuilder(StringTableBuilder::RAW, 1);
+  // creating external char pointer to store string builder table.
+  string charName = "__string_table_ptr";
+  LLVMTypeRef charPtrType = LLVMPointerType(LLVMInt8Type(), 0);
+  Constant *charValue = Constant::getNullValue(unwrap(charPtrType));
+  GlobalVariable *charGloablVar = new GlobalVariable(
+      *unwrap(modRef), unwrap(charPtrType), false, GlobalValue::ExternalLinkage,
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmPackageGen.cpp]: Changed to InternalLinkage.

File: compiler/BIR.h [hunk 12]
```diff
@@ -391,6 +391,8 @@ class TypeCastInsn : public NonTerminatorInsn {
   Operand *rhsOp;
   TypeDecl *typeDecl;
   bool checkTypes;
+  LLVMValueRef getIsSameTypeDeclaration (LLVMModuleRef &modRef,
```
Comment (KavinduZoysa) [inline compiler/BIR.h:394]: There is a space between function name and (. Please check other places too.

File: test/array/simpleArrayDeclaration.bal [hunk 13]
```diff
@@ -0,0 +1,9 @@
+// RUN: JAVA_HOME=%java_path %testRunScript %s %nballerinacc | filecheck %s
+
+int _bal_result = 0;
+public function main() {
+  int[] arr1 = [5];
+  _bal_result = 1;
+}
+
+// CHECK: RETVAL=1
```
Comment (KavinduZoysa) [inline test/array/simpleArrayDeclaration.bal:9]: Shall we add an example like this, 
```
public function main() {
	int[] a = [5];
	any x = 2;
	a[0] = <int>x;
}
```

File: compiler/llvmTypeCastInsn.cpp [hunk 14]
```diff
@@ -22,23 +22,60 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
...
-      //  If it returns false, then should be a call to abort().
 
+      // GEP of last type of smart pointer(original type of any variable(smart
+      // pointer))
+      LLVMValueRef lastTypeIdx =
+          LLVMBuildStructGEP(builder, rhsOpRef, 1, "lastTypeIdx");
+      LLVMValueRef lastTypeLoad = LLVMBuildLoad(builder, lastTypeIdx, "");
+      // sign extent of GEP to i64.
+      LLVMValueRef sExt =
+          LLVMBuildSExt(builder, lastTypeLoad, LLVMInt64Type(), "");
```
Comment (KavinduZoysa) [inline compiler/llvmTypeCastInsn.cpp:43]: @ramakotareddy , could you please explain the reason to do this, just for my knowledge.

Comment (KavinduZoysa) [inline compiler/llvmTypeCastInsn.cpp:43]: @ramakotareddy , could you please explain the reason to do this, just for my knowledge.

File: test/array/simpleArrayDeclaration.bal (hunk 13 above)
Comment (ramakotareddy) [reply test/array/simpleArrayDeclaration.bal:9]: Added inside test/array directory. test case name is "anyTypeVarAssignToArray.bal".

File: compiler/BIR.h (hunk 12 above)
Comment (ramakotareddy) [reply compiler/BIR.h]: Fixed

File: compiler/llvmFunctionGen.cpp [hunk 15]
```diff
@@ -149,8 +165,75 @@ void BIRFunction::translateFunctionBody(LLVMModuleRef &modRef) {
...
+// splitBB Basicblock (ifBB) and abortBB(elseBB).
+// In IfBB we are doing casing and from ElseBB Aborting.
+void BIRFunction::splitBBIfPossible (LLVMModuleRef &modRef) {
+  unsigned totalBB = LLVMCountBasicBlocks(getNewFunctionRef());
+  LLVMBasicBlockRef currentBB = LLVMGetFirstBasicBlock(getNewFunctionRef());
+  for (unsigned i = 0; i < totalBB; i++) {
+    BasicBlock *bBlock = unwrap(currentBB);
+    for (BasicBlock::iterator I = bBlock->begin(); I != bBlock->end(); ++I) {
+      CallInst *callInst = dyn_cast<CallInst>(&*I);
+      if (callInst) {
```
Comment (Kishanthan) [inline compiler/llvmFunctionGen.cpp:179]: Inverse this if and continue ?

File: test/array/anyTypeVarAssignToArray.bal [hunk 16]
```diff
@@ -0,0 +1,10 @@
+// RUN: JAVA_HOME=%java_path %testRunScript %s %nballerinacc | filecheck %s
```
Comment (Kishanthan) [inline test/array/anyTypeVarAssignToArray.bal:1]: Format all the Ballerina source files.

File: compiler/llvmTypeCastInsn.cpp [hunk 17]
```diff
@@ -92,3 +132,74 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
     }
   }
 }
+
+LLVMValueRef TypeCastInsn::getIsSameTypeDeclaration(LLVMModuleRef &modRef,
```
Comment (Kishanthan) [inline compiler/llvmTypeCastInsn.cpp:136]: All new additions in cpp files have to be formatted properly.

File: compiler/llvmFunctionGen.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmFunctionGen.cpp]: Fixed

File: compiler/llvmFunctionGen.cpp [hunk 18]
```diff
@@ -147,6 +162,71 @@ void BIRFunction::translateFunctionBody(LLVMModuleRef &modRef) {
...
+          LLVMValueRef brInsn = LLVMBuildBr(builder, elseBBSucc);
+          LLVMInstructionRemoveFromParent(brInsn);
+          // generate LLVMFunction call to Abort from elseLLVMBB(abortBB).
+          LLVMValueRef abortInsn = generateAbortInsn(builder, modRef);
+          LLVMInstructionRemoveFromParent(abortInsn);
+          // Inserting Abort Functioncall instruction into elseLLVMBB(abortBB).
+          unwrap(elseLLVMBB)
+              ->getInstList()
+              .push_back(unwrap<Instruction>(abortInsn));
+          unwrap(elseLLVMBB)
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmFunctionGen.cpp]: Fixed

File: compiler/llvmPackageGen.cpp [hunk 19]
```diff
@@ -27,6 +27,16 @@ LLVMValueRef BIRPackage::getGlobalVarRefUsingId(string globVar) {
 void BIRPackage::translate(LLVMModuleRef &modRef) {
   // String Table initialization
   strBuilder = new StringTableBuilder(StringTableBuilder::RAW, 1);
+  // creating external char pointer to store string builder table.
+  string charName = "__string_table_ptr";
```
[the remark this answers is not in the document]
Comment (ramakotareddy) [reply compiler/llvmPackageGen.cpp]: Fixed

File: compiler/llvmArrayTypeInsn.cpp (hunk 1 above)
Comment (ramakotareddy) [reply compiler/llvmArrayTypeInsn.cpp:31]: Fixed

File: compiler/llvmTypeCastInsn.cpp (hunk 17 above)
Comment (ramakotareddy) [reply compiler/llvmTypeCastInsn.cpp]: Fixed

File: compiler/llvmFunctionGen.cpp (hunk 15 above)
Comment (ramakotareddy) [reply compiler/llvmFunctionGen.cpp]: Fixed

File: compiler/llvmTypeCastInsn.cpp (hunk 4 above)
Comment (ramakotareddy) [reply compiler/llvmTypeCastInsn.cpp]: Fixed

File: test/array/anyTypeVarAssignToArray.bal (hunk 16 above)
Comment (ramakotareddy) [reply test/array/anyTypeVarAssignToArray.bal:1]: Formatted.

**Review by Kishanthan — COMMENTED** [review]
Please check the git commit guide we follow - https://chris.beams.io/posts/git-commit/

Please update the commit messages accordingly and squash related commits together.

**Comment by ramakotareddy** [comment]
Opened new PR #159

**Pull request closed without merging**